### PR TITLE
fix(TypeScript): merge now infers for N arguments

### PR DESCRIPTION
**Pull request by benlesh**
- infers for N arguments of type `ObservableInput`.
- infers for up to 6 arguments if `concurrency` or `scheduler` are passed (but those are deprecated)
- Otherwise returns `Observable<any>`.
- Adds `mergeConcurrent`
- Deprecates all versions of `merge` that accept a concurrency argument (in favor of `mergeConcurrent`)

High level:

This is the start of splitting `merge` into `merge` and `mergeConcurrent`, where the former is the most common use-case and the latter is the version that accepts a concurrency limit. The goal here is to provide better code readability and better typing.

Todo:

- [ ] Add tests for `mergeConcurrent`
- [ ] Add dtslint tests for `mergeConcurrent`
- [ ] Split commits into `feat` and `fix` commits.

### CODE REVIEW

**Comment by benlesh** [comment]
I suspect this might fail in CI because it's configured to use 2.8. We should bump that to 3.0 as well.

**Comment by kwonoj** [comment]
for same crux with https://github.com/ReactiveX/rxjs/pull/4405, if we bump up tsc into 3 shouldn't it be considered as breaking changes for v6?

**Comment by benlesh** [comment]
@kwonoj :thinking: yeah, possibly... maybe we just release v7 to include the typings? Type issues are a big problem for updating RxJS, and I'm sure most of our consumers are on 3.0+. It's probably time we stop coveting the major version as something special and start using it for what it's for. We don't need to remove deprecations, but at the very least we need to get these type changes in.

There's still a lot of work to do in this area though.

**Comment by kwonoj** [comment]
>  I'm sure most of our consumers are on 3.0

This is also my guess but we can't say for sure. What's ng's LTS's tsc version? do you have visibility?

+ I agree this is somewhat grey area and I would like to explicitly specify supported TS versions from v7 major.

**Comment by cartant** [comment]
> maybe we just release v7 to include the typings?

Releasing V7 with the improved typings means that `experimental` will be V8 and - with the perf improvements, that seems fitting 🏎

File: spec-dtslint/observables/merge-spec.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    it('should accept 6 arguments', () => {
+      const o = merge(a, b, c, d, e, f, 1); // $ExpectType Observable<A | B | C | D | E | F>
+    });
+
+    it('should accept more than 6 arguments', () => {
+      const o = merge(a, b, c, d, e, f, g, 1); // $ExpectType Observable<any>
+    });
+
+    it('should accept a rest of similar arguments', () => {
+      const o = merge(...[a, a, a, a, a, a, a, a, a, a, a, a], 1); // $ExpectType Observable<any>
```
Comment (cartant) [inline spec-dtslint/observables/merge-spec.ts]: When discussing the `pipe` typings with Rado, I think the consensus was that `{}` should be returned in these situations - rather then `any`. See [this comment](https://github.com/ReactiveX/rxjs/pull/3945#discussion_r205219721)

**Comment by benlesh** [comment]
@kwonoj Angular 6+ is on TypeScript 2.7+... however, google3 is on TypeScript 3.1 and Angular Master (soon to be released at version 8) is on TypeScript 3.2+.

That said, people still using Angular 6 could stay with RxJS 6 without issue. If we do this as a major release, it will be an incremental step, mostly around typings, and definitely for the better for most users.

**Comment by kwonoj** [comment]
I feel we can declare v6 TS supported for 2.x-ish, and call next major supports 3+ then? as you mentioned, this'll guarantee current user won't be blocked by sudden breaking bump with their compilers.

File: spec-dtslint/observables/merge-spec.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
+// TypeScript Version: 3.2
```
[the remark this answers is not in the document]
Comment (benlesh) [reply spec-dtslint/observables/merge-spec.ts]: I think you're right. That's cruft from fiddling with getting it to work.

**Comment by benlesh** [comment]
this is blocked because it'll have to be in the next major version

**Comment by benlesh** [comment]
I made some changes to this PR, see original post at the top.

**Comment by kolodny** [comment]
Note that this fails when using a generic for merge

```ts
it('should accept a rest of arguments with a generic', () => {
  const o = merge<any>(...[a, b, c]); // Error: Expected 2-3 arguments, but got 0 or more.
});
```

it's attempting to use this definition:

```ts
export declare function merge<T>(v1: ObservableInput<T>, scheduler: SchedulerLike): Observable<T>;
```

**Comment by cartant** [comment]
@kolodny Essentially, yours is the same problem as in https://github.com/ReactiveX/rxjs/issues/3989

**Comment by kolodny** [comment]
@cartant I don't think so because there is an overload where my code works, it just that by providing a single generic parameter to `merge`, we're forcing TS to choose an overload that doesn't support spreading rest args

**Comment by cartant** [comment]
I see. Why do you need to specify the type parameter explicitly in the first place?

**Comment by kolodny** [comment]
To specify the observable type, I guess. Is it better to cast the result instead?

**Comment by cartant** [comment]
@kolodny 

> Is it better to cast the result instead?

Yes. IMO, using explicit type parameters to effect type assertions should be avoided. I think that explicit type parameters in the value space - rather than the type space - are something to avoid wherever possible. I would suggest asserting either the argument or the return type, instead.

Explicit type parameters are unnecessary in situations in which TypeScript is able to make the correct inference. And in other situations, IMO, the assertion is better made closer to where the issue exists.

Also, when specifying an explicit type parameter, the code is more fragile, as it's exposed to any breaking changes in the typings.

TBH, I've felt - intuitively - that explicit type parameters in the value space are a 'bad thing' and have been attempting to come up with a succinct explanation since writing the [`rxjs-no-explicit-generics`](https://github.com/cartant/rxjs-tslint-rules/blob/master/source/rules/rxjsNoExplicitGenericsRule.ts) rule. However, the explanation above is the best that I have, ATM.

Also, I would suggest that this is a general TypeScript guideline - it's not something that's specific to RxJS.

**Comment by benlesh** [comment]
This is so old, I doubt it's of any value any longer.

**Pull request closed without merging**